app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,146 @@
+    }
+
+    /**
+     * Removes a folder from the cache model.
+     */
+    public void removeFolder(String folderId) {
+        for (String deviceId : deviceFolderMap.keySet()) {
+            if (deviceFolderMap.get(deviceId).containsKey(folderId)) {
+                deviceFolderMap.get(deviceId).remove(folderId);
+                break;
+            }
+        }
+    }
+
+    /**
+     * Updates device information in the cache model.
+     */
+    public void updateDevices(List<Device> newDeviceSet) {
+        List<String> removedDevices = new ArrayList<>();;
+        Boolean deviceFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            deviceFound = false;
+            for (Device device : newDeviceSet) {
+                if (device.deviceID.equals(deviceId)) {
+                    deviceFound = true;
+                    break;
+                }
+            }
+            if (!deviceFound) {
+                removedDevices.add(deviceId);
+            }
+        }
+        for (String deviceId : removedDevices) {
+            Log.v(TAG, "updateDevices: Remove device '" + deviceId + "' from cache model");
+            removeDevice(deviceId);
+        }
+    }
+
+    /**
+     * Updates folder information in the cache model.
+     */
+    public void updateFolders(List<Folder> newFolderSet) {
+        List<String> removedFolders = new ArrayList<>();;
+        Boolean folderFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            for (String folderId : deviceFolderMap.get(deviceId).keySet()) {
+                folderFound = false;
+                for (Folder folder : newFolderSet) {
+                    if (folder.id.equals(folderId)) {
+                        folderFound = true;
+                        break;
+                    }
+                }
+                if (!folderFound) {
+                    removedFolders.add(folderId);
+                }
+            }
+        }
+        for (String folderId : removedFolders) {
+            Log.v(TAG, "updateFolders: Remove folder '" + folderId + "' from cache model");
+            removeFolder(folderId);
+        }
+    }
+
+    /**
+     * Calculates remote device sync completion percentage across all folders
+     * shared with the device.
+     */
+    public int getDeviceCompletion(String deviceId) {
+        addDevice(deviceId);
+        int folderCount = 0;
+        double sumCompletion = 0;
+        for (String folderId : deviceFolderMap.get(deviceId).keySet()) {
+            sumCompletion += getCompletionInfo(deviceId, folderId).completion;
+            folderCount++;
+        }
+        return (int) Math.floor(sumCompletion / folderCount);
+    }
+
+    /**
+     * Returns completionInfo from the completion[deviceId][folderId] model.
+     */
+    public CompletionInfo getCompletionInfo(String deviceId, String folderId) {
+        addFolder(deviceId, folderId);
+        return (deviceFolderMap.get(deviceId)).get(folderId);

```

Comment:
Same here. getOrDefault

Reply:
I removed "getCompletionInfo" as it is not called externally atm. This removes redundant checks in "getDeviceCompletion" where is moved the getCompletion code part with getOrDefault.